Mathlib/Algebra/Polynomial/RuleOfSigns.lean

```diff
@@ -0,0 +1,588 @@
+    (hc : P.nextCoeff < 0) :
+    ((X - C η) * P).eraseLead.SignVariations = ((X - C η) * P.eraseLead).SignVariations := by
+  obtain ⟨dp1, hdp1⟩ := Nat.exists_eq_add_one.mpr (natDegree_pos_of_nextCoeff_ne_zero hc.ne)
+  have hndxP : natDegree ((X - C η) * P) = P.natDegree + 1 := by
+    have hPn0 : P ≠ 0 :=
+      leadingCoeff_ne_zero.mp hP₀.ne'
+    rw [natDegree_mul (X_sub_C_ne_zero η) hPn0, natDegree_X_sub_C, add_comm]
+  have hndxeP : natDegree ((X - C η) * P.eraseLead) = P.natDegree := by
+    have hePn0 : P.eraseLead ≠ 0 :=
+      mt nextCoeff_eq_zero_of_eraseLead_eq_zero hc.ne
+    rw [natDegree_mul (X_sub_C_ne_zero η) hePn0, natDegree_X_sub_C, add_comm]
+    exact natDegree_eraseLead_add_one hc.ne
+  have hQ : ((X - C η) * P).nextCoeff = coeff P dp1 - η * coeff P (dp1 + 1) := by
+    rw [nextCoeff_of_natDegree_pos (hndxP ▸ P.natDegree.succ_pos)]
+    rw [hndxP, hdp1, Nat.add_sub_cancel, coeff_X_sub_C_mul]
+  have hQ₁ : ((X - C η) * P).nextCoeff < 0 := by
+    rw [hQ, sub_neg]
+    rw [nextCoeff_of_natDegree_pos (hdp1 ▸ dp1.succ_pos), hdp1, Nat.add_sub_cancel] at hc
+    rw [leadingCoeff, hdp1] at hP₀
+    exact hc.trans (mul_pos hη hP₀)
+  have hndexP0 : natDegree (eraseLead ((X - C η) * P)) = P.natDegree := by
+    apply Nat.add_right_cancel (m := 1)
+    rw [← hndxP, natDegree_eraseLead_add_one hQ₁.ne]
+  --the theorem is true mainly because all the signs are the same;
+  --in fact, the coefficients are all the same except the first.
+  suffices eraseLead (eraseLead ((X - C η) * P)) = eraseLead ((X - C η) * P.eraseLead) by
+    suffices (coeffList (eraseLead ((X - C η) * P))).map SignType.sign =
+      (coeffList ((X - C η) * P.eraseLead)).map SignType.sign by
+        rw [SignVariations, SignVariations, this]
+    rw [coeffList_eraseLead (mt nextCoeff_eq_zero_of_eraseLead_eq_zero hQ₁.ne)]
+    have hndeP0 : 0 < natDegree ((X - C η) * P.eraseLead) :=
+      Nat.lt_of_sub_eq_succ (hdp1 ▸ hndxeP)
+    rw [coeffList_eraseLead (ne_zero_of_natDegree_gt hndeP0)]
+    rw [List.map_cons, List.map_cons, leadingCoeff_eraseLead_eq_nextCoeff hQ₁.ne]
+    rw [sign_neg hQ₁, sign_neg ?_, this, hndxeP, hndexP0]
+    rwa [leadingCoeff_mul, leadingCoeff_X_sub_C, one_mul, leadingCoeff_eraseLead_eq_nextCoeff hc.ne]
+  rw [← self_sub_monomial_natDegree_leadingCoeff, leadingCoeff_eraseLead_eq_nextCoeff hQ₁.ne]
+  rw [hndexP0]
+  rw [← self_sub_monomial_natDegree_leadingCoeff, leadingCoeff_monic_mul (monic_X_sub_C η)]
+  rw [← self_sub_monomial_natDegree_leadingCoeff, leadingCoeff_monic_mul (monic_X_sub_C η)]
+  rw [hndxeP, hndxP]
+  rw [leadingCoeff_eraseLead_eq_nextCoeff hc.ne, ← self_sub_monomial_natDegree_leadingCoeff]
+  rw [hQ, mul_sub, sub_mul, sub_mul, X_mul_monomial, C_mul_monomial, monomial_sub]
+  rw [leadingCoeff, nextCoeff_of_natDegree_pos (hdp1 ▸ dp1.succ_pos), hdp1, Nat.add_sub_cancel]
+  abel
+
+--effectively a specialization of the final theorem for monomials
```

Comment:
Can this, and the other -- comments above lemmas in this file, be turned into a docstring comment for the lemma s?